maptivist/src/test/java/com/google/sps/MarkerServletFunctionTest.java

```diff
@@ -0,0 +1,104 @@
+
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+ 
+package com.google.sps.servlets;
+ 
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import java.util.Base64;
+ 
+ 
+/** */
+@RunWith(JUnit4.class)
+public final class MarkerServletFunctionTest {
+ 
+    private static final String[] categoryArray = new String[] {"BLM", "LGBTQ"};
+    private static final String[] flagArray = new String[] {"boy", "girl", "cow"};
+    private static final String[] linkArray = new String[] {"www.google.com", "www.youtube.com", "www.pinterest.com"};
+ 
+    private static final Set<String> categoryObject = new HashSet<String>(Arrays.asList(categoryArray));
+    private static final ArrayList<String> flagObject = new ArrayList<String>(Arrays.asList(flagArray));
+    private static final Set<String> linkObject = new HashSet<String>(Arrays.asList(linkArray));
+ 
+    Base64.Decoder decoder = Base64.getDecoder();  
+    Base64.Encoder encoder = Base64.getEncoder();  
+ 
+ 
+    private MarkerServlet testMarkerServletFunctions;
+ 
+    @Before
+    public void setUp() {
+        testMarkerServletFunctions = new MarkerServlet();
+    }
+ 
+    @Test
+    public void testCreateCategoriesString() {
+        String actual = testMarkerServletFunctions.createCategoriesString(categoryObject);
+        String expected = "QkxN,TEdCVFE=";
```

Comment:
For these tests, think carefully about what you want to test and try to only test that.

For instance, you don't really care what the encoder does, as long as:
1) It doesn't mess with using the comma character is a string delimiter.
2) That you can encode and then decode the values.

So, what I suggest:

Combine testCreateCategoriesString() and testCreateCategoriesObject

for 1) Encode the categoryObject, and then ensure that it only has 1 comma (since there are two entries). When testing the linkArray there should be 2 commas, etc. Also, it's a good idea to put commas in the test strings to make sure they get encoded properly. Don't bother looking at any other part of the encoded string.

for 2) Take the encoded string and call createCategoriesObject and verify that it matches the expected categoryObject.

LMK if that makes sense.